### DISCUSSION TITLE
SQLModel seems to disable @model_validator when initialized with table=true

### DISCUSSION THREAD

**Discussion by eric-pa** [fastapi/sqlmodel#805] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlmodel import SQLModel, Field
from pydantic import model_validator

class SQLModelTest(SQLModel, table=True):
    id: str = Field(primary_key=True)

    @model_validator(mode="before")
    @classmethod
    def test(cls, data):
        print("Hello from sqlmodel")
        return data

sqlmodel_test = SQLModelTest(id="123")
```

### Description

Hi,

I'm having problems migrating to pydantic v2. I have some logic in @root_validator for data that should be stored in the database. However, when using pydantic v2 and SQLModel with `table=True` the new @model_validator is not called. Everything works fine when `table=False`.

Is this a bug or am I doing something wrong?

Best
Eric

### Operating System

Linux

### SQLModel Version

0.0.14

### Python Version

Python 3.12.1

**Comment by rt87** [comment]
Same for me (with version 0.0.22)

**Reply by monchin** [reply]
I also found this bug and made a PR to fix it in #1041 but no responses so far.

For now, I found that `your_obj = YourModel.model_validate(your_dict)` would enable pydantic validation. In this discussion, it would be `sqlmodel_test = SQLModelTest.model_validate({"id": "123"})`

**Reply by mg3146** [reply]
I've tried this, but I'm getting "object has no attribute '_sa_instance_state'" errors in some of my model_validators

**Reply by iloveitaly** [reply]
Has anyone found a workaround here?

**Reply by monchin** [reply]
> but I'm getting "object has no attribute '_sa_instance_state'" errors in some of my model_validators

Would you offer an example to reproduce it?

**Reply by sr-verde** [reply]
I have the same problem. 

```
AttributeError: 'Domain' object has no attribute '_sa_instance_state'
```

This would be an example for it:

```python
class DomainBase(SQLModel):
    domain: str = Field(unique=True, index=True)

class Domain(DomainBase, table=True):
    id: int | None = Field(default=None, primary_key=True)
    hashed_domain: str | None = Field(default=None, unique=True, index=True)

    @model_validator(mode="after")
    def set_hashed_domain(self) -> Self:
        self.hashed_domain = "hashed"

        return self
```

The issue will disappear when you remove `self.hashed_domain = "hashed"`. My workaround is to use a `mode="before"`. For me, it’s okay but that may not work for other use cases.

**Comment by MattOates** [comment]
Not sure if this is still an issue or not, but I found this searching for the same problem. We found this is related to some kind of lazy evaluation of the validator. If you edit any one member of the model before inspecting it then you do see the model level validation happening. Feels relatively bug like as you'd expect at init time or shortly after for this to happen and not require any further editing of data. Hope this helps someone.
